File: php/tests/ClinicalExampleTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\Registry;
use PHPUnit\Framework\TestCase;

use function Oojs\validate;

/**
 * Integration tests using the clinical example schema and instances.
 * Reuses the JSON fixtures from ../../examples/.
 */
class ClinicalExampleTest extends TestCase
{
    use TestHelpers;

    private Registry $r;
    private \Oojs\Model\Schema $schema;
    private array $validInstances;
    private array $invalidInstances;

    protected function setUp(): void
    {
        $examples = __DIR__ . '/../../examples';
        $this->r  = new Registry();
        $this->schema = $this->r->loadFile("$examples/clinical.oojs.json");

        $data = json_decode(
            file_get_contents("$examples/clinical-instances.json"),
            associative: true,
            flags: JSON_THROW_ON_ERROR,
        );
        $this->validInstances   = $data['valid'];
        $this->invalidInstances = $data['invalid'];
    }

    public function test_valid_instances_pass(): void
    {
        foreach ($this->validInstances as $inst) {
            $clean = array_filter($inst, static fn($k) => $k !== '_comment', ARRAY_FILTER_USE_KEY);
            $dv    = $clean['_type'] ?? null;
            if ($dv === null || !isset($this->schema->types[$dv])) {
                continue;
            }
            $typedef = $this->schema->types[$dv];
            $errs    = validate($clean, $typedef, $this->schema, $this->r);
            $this->assertSame(
                [],
                $errs,
                "Expected valid but got errors for $dv: " . implode(', ', array_map('strval', $errs)),
            );
        }
    }

    public function test_invalid_instances_fail(): void
    {
        foreach ($this->invalidInstances as $inst) {
            $dv = $inst['_type'] ?? null;
            if ($dv === null) {
                $target = $this->schema->types['Observation'];
            } else {
                $target = $this->schema->types[$dv]
                    ?? array_values($this->schema->types)[0];
            }
            $errs    = validate($inst, $target, $this->schema, $this->r);
            $comment = $inst['_comment'] ?? '';
            $this->assertNotEmpty($errs, "Expected errors for invalid instance ($comment)");
        }
    }
}

File: java/src/test/java/org/oojs/ClinicalExampleTest.java
package org.oojs;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;
import org.oojs.model.TypeDef;

import java.io.File;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class ClinicalExampleTest extends TestHelpers {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private Registry r;
    private Schema schema;
    private List<Map<String, Object>> validInstances;
    private List<Map<String, Object>> invalidInstances;

    @BeforeEach
    @SuppressWarnings("unchecked")
    void setUp() throws Exception {
        String examples = Paths.get(System.getProperty("user.dir"))
                .getParent().resolve("examples").toString();
        r = new Registry();
        schema = r.loadFile(examples + "/clinical.oojs.json");

        Map<String, Object> data = MAPPER.readValue(
                new File(examples + "/clinical-instances.json"), Map.class);
        validInstances = (List<Map<String, Object>>) data.get("valid");
        invalidInstances = (List<Map<String, Object>>) data.get("invalid");
    }

    @Test void valid_instances_pass() {
        for (Map<String, Object> inst : validInstances) {
            Map<String, Object> clean = new java.util.LinkedHashMap<>(inst);
            clean.remove("_comment");
            String dv = (String) clean.get("_type");
            if (dv == null || !schema.types.containsKey(dv)) continue;
            TypeDef typedef = schema.types.get(dv);
            List<ValidationError> errs = new Validator(r).validate(clean, typedef, schema);
            assertTrue(errs.isEmpty(), "Expected valid but got errors for " + dv + ": " + errs);
        }
    }

    @Test void invalid_instances_fail() {
        for (Map<String, Object> inst : invalidInstances) {
            String dv = (String) inst.get("_type");
            TypeDef target;
            if (dv == null) {
                target = schema.types.values().iterator().next();
            } else {
                target = schema.types.getOrDefault(dv, schema.types.values().iterator().next());
            }
            List<ValidationError> errs = new Validator(r).validate(inst, target, schema);
            String comment = (String) inst.getOrDefault("_comment", "");
            assertFalse(errs.isEmpty(), "Expected errors for invalid instance (" + comment + ")");
        }
    }
}
